smack-experimental/src/main/java/org/jivesoftware/smackx/reference/element/ReferenceElement.java

```diff
@@ -0,0 +1,141 @@
+    private final Integer end;
+    private final Type type;
+    private final String anchor;
+    private final String uri;
+
+    // Non-XEP-compliant, but needed for SIMS
+    private final ExtensionElement child;
+
+    /**
+     * XEP-incompliant (v0.2) constructor. This is needed for SIMS.
+     *
+     * @param begin
+     * @param end
+     * @param type
+     * @param anchor
+     * @param uri
+     * @param child
+     */
+    public ReferenceElement(Integer begin, Integer end, Type type, String anchor, String uri, ExtensionElement child) {
+        if (begin != null && begin < 0) {
+            throw new IllegalArgumentException("Attribute 'begin' MUST NOT be smaller than 0.");
+        }
+        if (end != null && end < 0) {
+            throw new IllegalArgumentException("Attribute 'end' MUST NOT be smaller than 0.");
+        }
+        if (begin != null && end != null && begin >= end) {
+            throw new IllegalArgumentException("Attribute 'begin' MUST be smaller than attribute 'end'.");
+        }
+        if (type == null) {
+            throw new NullPointerException("Attribute 'type' MUST NOT be null.");
+        }
+        // TODO: The uri attribute is not mandatory according to SIMS, but it is according to references.
+        /*if (uri == null) {
+            throw new NullPointerException("Attribute 'uri' MUST NOT be null.");
+        }*/
+        this.begin = begin;
+        this.end = end;
+        this.type = type;
+        this.anchor = anchor;
+        this.uri = uri;
+        this.child = child;
+    }
+
+    /**
+     * XEP-Compliant constructor.
+     *
+     * @param begin
+     * @param end
+     * @param type
+     * @param anchor
+     * @param uri
+     */
+    public ReferenceElement(Integer begin, Integer end, Type type, String anchor, String uri) {
+        this(begin, end, type, anchor, uri, null);
+    }
+
+    public Integer getBegin() {
+        return begin;
+    }
+
+    public Integer getEnd() {
+        return end;
+    }
+
+    public Type getType() {
+        return type;
+    }
+
+    public String getAnchor() {
+        return anchor;
+    }
+
+    public String getUri() {
+        return uri;
+    }
+
+    @Override
+    public String getNamespace() {
+        return ReferenceManager.NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    @Override
+    public XmlStringBuilder toXML() {
+        XmlStringBuilder xml = new XmlStringBuilder(this)
+                .optIntAttribute(ATTR_BEGIN, begin != null ? begin : -1)
+                .optIntAttribute(ATTR_END, end != null ? end : -1)
+                .attribute(ATTR_TYPE, type.toString())
+                .optAttribute(ATTR_ANCHOR, anchor)
+                .optAttribute(ATTR_URI, uri);
+
+        if (child == null) {
+            return xml.closeEmptyElement();
+        } else {
+            return xml.rightAngleBracket()
+                    .append(child.toXML())
```

Comment:
I think just `.append(child)` should be sufficient.

Reply:
Nope, the argument of `append()` cannot be an `ExtensionElement`.